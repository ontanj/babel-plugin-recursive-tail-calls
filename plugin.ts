import { type NodePath, types } from "@babel/core";
import type {
  CallExpression,
  Expression,
  Function,
  Identifier,
} from "@babel/types";

type t = typeof types;

interface State {
  // true if we should apply the transformation for this function
  recursion: boolean;
  // loop label
  labelIdentifier: Identifier;
  // identifier of this function
  functionIdentifier: Identifier;
  // identifier of loop condition
  conditionIdentifier: Identifier;
  // path of this function
  functionPath: NodePath<Function>;
  // name and default value of function argument
  arguments: { identifier: Identifier; defaultValue: Expression }[];
}

export default function ({ types: t }: { types: t }) {
  const callExpVisitor = {
    CallExpression(this: State, path: NodePath<CallExpression>) {
      const callsItself =
        t.isIdentifier(path.node.callee) &&
        path.scope.bindingIdentifierEquals(
          path.node.callee.name,
          this.functionIdentifier,
        );
      const isLast = t.isReturnStatement(path.parent);
      const shouldOptimize = callsItself && isLast;

      if (!shouldOptimize) return;

      this.recursion = true;

      const args = this.arguments.map(({ identifier, defaultValue }, index) => {
        return {
          identifier,
          value: path.node.arguments[index] ?? defaultValue,
        };
      });

      const updateExpression = t.expressionStatement(
        t.assignmentExpression(
          "=",
          t.arrayPattern(args.map(({ identifier }) => identifier)),
          t.arrayExpression(args.map(({ value }) => value as Expression)),
        ),
      );

      // the parent is ReturnStatement
      path.parentPath.insertBefore(updateExpression);
      path.parentPath.insertBefore(
        t.expressionStatement(
          t.assignmentExpression(
            "=",
            this.conditionIdentifier,
            t.booleanLiteral(true),
          ),
        ),
      );
      path.parentPath.insertBefore(t.continueStatement(this.labelIdentifier));
      path.parentPath.remove();
    },

    Function(path: NodePath<Function>) {
      // skip nested functions
      path.skip();
    },
  };

  return {
    visitor: {
      Function(path: NodePath<Function>) {
        const functionIdentifier = getFunctionIdentifier(path, t);
        if (!functionIdentifier) return;

        const functionBody = path.get("body");
        // until we support ternary, we can't have expression body
        if (!functionBody.isBlockStatement()) return;

        const labelIdentifier =
          path.scope.generateUidIdentifier("tail-call-loop");

        const conditionIdentifier =
          path.scope.generateUidIdentifier("continue-recursion");

        let args: State["arguments"];

        try {
          args = path.node.params.map(
            (param: (typeof path.node.params)[number]) => {
              if (t.isIdentifier(param)) {
                return {
                  identifier: param,
                  defaultValue: t.identifier("undefined"),
                };
              } else if (
                t.isAssignmentPattern(param) &&
                t.isIdentifier(param.left)
              ) {
                return { identifier: param.left, defaultValue: param.right };
              }
              throw new Error("Unsupported param expression");
            },
          );
        } catch (e: unknown) {
          return;
        }

        const state: State = {
          recursion: false,
          labelIdentifier,
          functionIdentifier,
          conditionIdentifier,
          functionPath: path,
          arguments: args,
        };

        path.traverse(callExpVisitor, state);

        // abort if there is no recursion
        if (!state.recursion) return;

        const conditionDeclaration = t.variableDeclaration("let", [
          t.variableDeclarator(conditionIdentifier, t.booleanLiteral(true)),
        ]);

        // wrap function body in while loop
        const whileStatement = t.whileStatement(
          conditionIdentifier,
          functionBody.node,
        );
        // insert `condition = false` first in loop
        functionBody.unshiftContainer(
          "body",
          t.expressionStatement(
            t.assignmentExpression(
              "=",
              conditionIdentifier,
              t.booleanLiteral(false),
            ),
          ),
        );
        const labeledStatement = t.labeledStatement(
          labelIdentifier,
          whileStatement,
        );
        const blockStatement = t.blockStatement([
          conditionDeclaration,
          labeledStatement,
        ]);

        functionBody.replaceWith(blockStatement);
      },
    },
  };
}

function getFunctionIdentifier(functionPath: NodePath<Function>, t: t) {
  if (t.isFunctionDeclaration(functionPath.node)) {
    return functionPath.node.id;
  } else if (t.isArrowFunctionExpression(functionPath.node)) {
    if (
      t.isVariableDeclarator(functionPath.parent) &&
      t.isIdentifier(functionPath.parent.id) &&
      functionPath.scope.getBinding(functionPath.parent.id.name)?.constant
    ) {
      return functionPath.parent.id;
    }
  }
}
