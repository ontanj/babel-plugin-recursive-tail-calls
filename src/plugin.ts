import { type NodePath, types } from "@babel/core";
import {
  isArrowFunctionExpression,
  isFunctionDeclaration,
  isIdentifier,
  isVariableDeclarator,
  type Function,
} from "@babel/types";
import {
  callExpressionRewriter,
  type State as CallExpressionState,
} from "./callExpressionRewriter.js";
import { findRecursion } from "./tailRecursionFinder.js";

type t = typeof types;

export default function ({ types: t }: { types: t }) {
  return {
    visitor: {
      Function(path: NodePath<Function>) {
        const functionIdentifier = getFunctionIdentifier(path);
        if (!functionIdentifier) return;

        const functionBody = path.get("body");

        const labelIdentifier =
          path.scope.generateUidIdentifier("tail-call-loop");

        const conditionIdentifier =
          path.scope.generateUidIdentifier("continue-recursion");

        let args: CallExpressionState["arguments"];

        try {
          args = path.node.params.map(
            (param: (typeof path.node.params)[number]) => {
              if (t.isIdentifier(param)) {
                return {
                  identifier: param,
                  defaultValue: t.identifier("undefined"),
                };
              } else if (
                t.isAssignmentPattern(param) &&
                t.isIdentifier(param.left)
              ) {
                return { identifier: param.left, defaultValue: param.right };
              }
              throw new Error("Unsupported param expression");
            },
          );
        } catch (e: unknown) {
          return;
        }

        if (functionBody.isExpression()) {
          if (!findRecursion(functionBody, functionIdentifier)) return;
          functionBody.replaceWith(
            t.blockStatement([t.returnStatement(functionBody.node)]),
          );
        }
        // this should never happen but is added for adequate type hints
        if (!functionBody.isBlockStatement())
          throw new Error("Body is not block statement");

        const state: CallExpressionState = {
          recursion: false,
          labelIdentifier,
          functionIdentifier,
          conditionIdentifier,
          functionPath: path,
          arguments: args,
        };

        path.traverse(callExpressionRewriter, state);

        // abort if there is no recursion
        if (!state.recursion) return;

        const conditionDeclaration = t.variableDeclaration("let", [
          t.variableDeclarator(conditionIdentifier, t.booleanLiteral(true)),
        ]);

        // wrap function body in while loop
        const whileStatement = t.whileStatement(
          conditionIdentifier,
          functionBody.node,
        );
        // insert `condition = false` first in loop
        functionBody.unshiftContainer(
          "body",
          t.expressionStatement(
            t.assignmentExpression(
              "=",
              conditionIdentifier,
              t.booleanLiteral(false),
            ),
          ),
        );
        const labeledStatement = t.labeledStatement(
          labelIdentifier,
          whileStatement,
        );
        const blockStatement = t.blockStatement([
          conditionDeclaration,
          labeledStatement,
        ]);

        functionBody.replaceWith(blockStatement);
      },
    },
  };
}

function getFunctionIdentifier(functionPath: NodePath<Function>) {
  if (functionPath.isFunctionDeclaration()) {
    return functionPath.node.id;
  } else if (functionPath.isArrowFunctionExpression()) {
    if (
      isVariableDeclarator(functionPath.parent) &&
      isIdentifier(functionPath.parent.id) &&
      functionPath.scope.getBinding(functionPath.parent.id.name)?.constant
    ) {
      return functionPath.parent.id;
    }
  }
}
