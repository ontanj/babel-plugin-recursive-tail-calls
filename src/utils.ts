import { NodePath } from "@babel/core";
import { CallExpression, isIdentifier } from "@babel/types";
import { Identifier } from "@babel/types";

export function isRecCall(
  path: NodePath<CallExpression>,
  identifier: Identifier,
) {
  return (
    isIdentifier(path.node.callee) &&
    path.scope.bindingIdentifierEquals(path.node.callee.name, identifier)
  );
}
