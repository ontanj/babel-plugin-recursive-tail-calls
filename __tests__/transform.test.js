import babel from '@babel/core';
import plugin from "..";
import {
  base,
  multipleArgs,
  defaultArgInactive,
  defaultArgActive,
  undefinedArg,
  arrowFunction
} from "../examples/transform";
import { mapToTestCases } from "./utils";

const cases = mapToTestCases({
  base,
  multipleArgs,
  defaultArgInactive,
  defaultArgActive,
  undefinedArg,
  arrowFunction
})

describe("transform", () => {
  it.each(cases)('%s', (_, testCase) => {
    const { code } = babel.transform(testCase, { plugins: [plugin] });
    expect(code).toMatchSnapshot();
  });
});

