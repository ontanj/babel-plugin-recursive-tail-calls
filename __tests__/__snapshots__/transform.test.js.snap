// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform arrowFunction 1`] = `
"const arrowFunction = a => {
  _tailCallLoop: while (true) {
    if (a <= 0) return a;
    a = a - 1;
    continue _tailCallLoop;
  }
};"
`;

exports[`transform base 1`] = `
"function base(a) {
  _tailCallLoop: while (true) {
    if (a <= 0) {
      return a;
    }
    a = a - 1;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform defaultArgActive 1`] = `
"function defaultArgActive(a, b = 0) {
  _tailCallLoop: while (true) {
    if (a < b) return a;
    a = a - 1;
    b = 0;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform defaultArgInactive 1`] = `
"function defaultArgInactive(a, b = 0) {
  _tailCallLoop: while (true) {
    if (a < b) return a;
    a = a - 1;
    b = b + 1;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform multipleArgs 1`] = `
"function multipleArgs(a, b) {
  _tailCallLoop: while (true) {
    if (a < b) return a;
    a = a - 1;
    b = b + 1;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform multipleReturns 1`] = `
"function multipleReturns(a, b, c) {
  _tailCallLoop: while (true) {
    if (a < 0) return a;
    if (b < 0) {
      const n = b + c;
      a = n;
      b = b;
      c = c + 1;
      continue _tailCallLoop;
    }
    a = a + 1;
    b = b + 1;
    c = c;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform nestedStatements 1`] = `
"function nestedStatements(a, b) {
  _tailCallLoop: while (true) {
    if (a > b) {
      if (a > 0) {
        a = a - 1;
        b = b;
        continue _tailCallLoop;
      }
      return a - b;
    }
    return a + b;
  }
}"
`;

exports[`transform undefinedArg 1`] = `
"function undefinedArg(a, b) {
  _tailCallLoop: while (true) {
    if (a < b) return a;
    a = a - 1;
    b = undefined;
    continue _tailCallLoop;
  }
}"
`;
